### PR TITLE
feat(tianmu) Refactor the aggregation and The IN operator to multithread #472 (#422 #465 #466)

### LINKED ISSUE #422: feature:  make aggregation parallel process to reduce execution time
### Documentation for mysql aggregation

 https://dev.mysql.com/doc/refman/8.0/en/aggregate-functions-and-modifiers.html

## Functional Requirements:

1. Consistent with aggregation conditions handled by existing logic
2. The Aggregate function is used as the entry and ResultSender is used as the result output

### Table data limit:

1. The upper limit of a physical drive
2. The memory limit is exceeded. Procedure

### Memory usage limit:

1. Add parameters to control the maximum amount of memory that can be used for aggregation
2. Otherwise temporary disk files if the limit is exceeded

### Execution time limit:

1. Under the same conditions, it is in the same order of magnitude as InnoDB aggregate query at most
3. Aim to take less time than InnoDB for aggregated queries

 
##  Aggregate query  SQL

```sql

	select
		c_name,
		c_custkey,
		o_orderkey,
		o_orderdate,
		o_totalprice,
		sum(l_quantity)
	from
		customer,
		orders,
		lineitem
	where
		c_custkey = o_custkey
		and o_orderkey = l_orderkey
	group by
		c_name,
		c_custkey,
		o_orderkey,
		o_orderdate,
		o_totalprice
	order by
		o_totalprice desc,
		o_orderdate
	limit 10;

```

```sql
mysql> desc part;
+---------------+---------------+------+-----+---------+-------+
| Field         | Type          | Null | Key | Default | Extra |
+---------------+---------------+------+-----+---------+-------+
| p_partkey     | int(11)       | NO   | PRI | NULL    |       |
| p_name        | varchar(55)   | NO   |     | NULL    |       |
| p_mfgr        | char(25)      | NO   |     | NULL    |       |
| p_brand       | char(10)      | NO   |     | NULL    |       |
| p_type        | varchar(25)   | NO   |     | NULL    |       |
| p_size        | int(11)       | NO   |     | NULL    |       |
| p_container   | char(10)      | NO   |     | NULL    |       |
| p_retailprice | decimal(15,2) | NO   |     | NULL    |       |
| p_comment     | varchar(23)   | NO   |     | NULL    |       |
+---------------+---------------+------+-----+---------+-------+
9 rows in set (0.00 sec)


```

```sql

mysql> desc partsupp;
+---------------+---------------+------+-----+---------+-------+
| Field         | Type          | Null | Key | Default | Extra |
+---------------+---------------+------+-----+---------+-------+
| ps_partkey    | int(11)       | NO   | PRI | NULL    |       |
| ps_suppkey    | int(11)       | NO   | PRI | NULL    |       |
| ps_availqty   | int(11)       | NO   |     | NULL    |       |
| ps_supplycost | decimal(15,2) | NO   |     | NULL    |       |
| ps_comment    | varchar(199)  | NO   |     | NULL    |       |
+---------------+---------------+------+-----+---------+-------+
5 rows in set (0.00 sec)


```

```sql

    select
        p_brand,
        p_type,
        p_size,
        count(distinct ps_suppkey) as supplier_cnt
    from
        partsupp,
        part
    where
        p_partkey = ps_partkey
        and p_brand <> 'Brand#45'
        and p_type not like 'MEDIUM POLISHED%'
        and p_size in (49, 14, 23, 45, 19, 3, 36, 9)
        and ps_suppkey not in (
            select
                s_suppkey
            from
                supplier
            where
                s_comment like '%Customer%Complaints%'
        )
    group by
        p_brand,
        p_type,
        p_size
    order by
        supplier_cnt desc,
        p_brand,
        p_type,
        p_size
    limit 10;

```

### PR DESCRIPTION
**Pull request by adofsauron**
Issue Number: close #422  #465 #466

## Tests Check List

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

## Changelog

- [ ] New Feature
- [ ] Bug Fix
- [x] Improvement
- [ ] Performance Improvement
- [ ] Build/Testing/CI/CD
- [ ] Documentation
- [ ] Not for changelog (changelog entry is not required)

## Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features

### CODE REVIEW

**Comment by adofsauron** [comment]
​

Abstract:

Outline the design of aggregate multithreaded parallelism



https://github.com/stoneatom/stonedb/issues/422



Demand analysis:

Functional Requirements:

Consistent with aggregation conditions handled by existing logic

The Aggregate function is used as the entry and ResultSender is used as the result output

Performance Requirements:

Table data limit:

The upper limit of a physical drive

The memory limit is exceeded. Procedure

Memory usage limit:

Add a parameter that controls the maximum amount of memory that can be used during aggregation

A temporary disk file is used if the limit is exceeded

Execution time limit:

Under the same conditions, it is in the same order of magnitude as InnoDB aggregate query at most

Aim to take less time than InnoDB for aggregated queries

Summary Feature breakdown:

Partition of table tuples:

Segmentation target:

The table data traversed by a single thread is evenly divided into different intervals, so that different worker threads can independently process a certain interval of data

Correctly handle the memory visibility of different worker threads to ensure that the function of the worker thread itself is complete and correct, and will not cause different threads to work with each other due to the overlap of critical sections

Facing Problems:

How do I get the table tuple description, how many rows, how many columns, how many packs?

What are the rules for dividing a tablespace into intervals, how many worker threads are started, and how many intervals are handled by each worker thread?

GroupByWrapper currently stores the complete information of the table and holds the identity information of whether a row of the column has a block of data, which makes it impossible for multiple threads to process the class in parallel. Whether to maintain a GroupByWrapper class in each thread to store only the information about the current tuple range, so that the worker thread can handle it independently

The GroupByWrapper class stores the information of the matching aggregation column, which is the same in each worker thread. Whether to separate the information of the aggregation column into a separate data structure, each worker thread will read only

Should we also use the current encapsulation layers of GroupByWrapper and GroupTable, GroupDistinctTable and GroupDistinctCache, and does that really cut functionality correctly? What is the maintenance cost? Is it already the optimal class organization structure?

The ValueMatchingTable class, which previously stored aggregated data, should now be replaced with a Parallel Hash Map, which would otherwise require data to be stored in each worker thread. Finally, the data in each worker thread is copied together to do a global aggregate processing operation. The performance overhead of memory copy is huge and unacceptable. How does the interface of the Parallel Hash Map class combine with the logic of the upper-level traversal aggregation when replacing a local hash?

When writing aggregate data, how does a parallel Hash map perform aggregation operation with existing data to form a unified result? How to ensure the correctness of aggregate summary results?

What are the lock rules when a parallel Hash map involves multiple threads competing to write the value of the same aggregation condition? How can I avoid overwriting the results of different worker threads?

Strip the column attributes of the processing aggregate:

Process aggregated columns whose meta-attributes are independent of a particular tuple and are invariant and can be stripped

The GroupByWrapper class retains data associated with tuples during the tuple traversal, such as rows that have been processed, data that has been aggregated, etc











​

File: .jenkins/jenkinsfile-mysql8-centos7-30-119 [hunk 1]
```diff
@@ -0,0 +1,558 @@
+pipeline {
+    agent {
```
Comment (RingsC) [inline .jenkins/jenkinsfile-mysql8-centos7-30-119]: I suggest you committing these jenkins in another independent ISSUE and PR.

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 2]
```diff
@@ -194,7 +194,7 @@ void AggregationAlgorithm::Aggregate(bool just_distinct, int64_t &limit, int64_t
     }
   } else {
     int64_t local_limit = limit == -1 ? upper_approx_of_groups : limit;
-    MultiDimensionalGroupByScan(gbw, local_limit, offset, sender, limit_less_than_no_groups);
```
Comment (RingsC) [inline storage/tianmu/core/aggregation_algorithm.cpp:197]: Pls, follow the commit guide. The PR title, commit message, etc.

File: storage/tianmu/vc/virtual_column_base.cpp [hunk 3]
```diff
@@ -164,7 +164,6 @@ int64_t VirtualColumnBase::GetApproxDistVals(bool incl_nulls, core::RoughMultiIn
 
 int64_t VirtualColumnBase::GetMaxInt64(const core::MIIterator &mit) {
   int64_t res = GetMaxInt64Impl(mit);
-  DEBUG_ASSERT(res != common::NULL_VALUE_64);
```
Comment (RingsC) [inline storage/tianmu/vc/virtual_column_base.cpp]: Why do you delete this line? it's a protective code.

File: storage/tianmu/core/value_matching_table.h [hunk 4]
```diff
@@ -52,6 +52,10 @@ class ValueMatchingTable {  // abstract class: interface for value matching
   // already exists, false if put as a new row
   virtual bool FindCurrentRow(unsigned char *input_buffer, int64_t &row, bool add_if_new = true) = 0;
 
+  virtual bool FindCurrentRow(unsigned char *input_buffer, int64_t &row, bool add_if_new, int match_width) {
```
Comment (RingsC) [inline storage/tianmu/core/value_matching_table.h:55]: It's not A GOOD IDEA that adds these codes here. 
```
 37 class ValueMatchingTable {  // abstract class: interface for value matching
 38                             // (Group By etc.)

```
and, what's the meaning of your input params, `match_width` ?

Comment (RingsC) [reply storage/tianmu/core/value_matching_table.h:55]: BTW: your  can make this function to be  readability.
```
 int64_t NoRows() { return int64_t(no_rows); }  // rows stored so far
        ..
 virtual bool NoMoreSpace() = 0;                // hard stop: any new value will be rejected
```

File: storage/tianmu/core/value_matching_table.cpp [hunk 5]
```diff
@@ -52,11 +52,13 @@ ValueMatchingTable *ValueMatchingTable::CreateNew_ValueMatchingTable(int64_t mem
   }
 
   // easy case: narrow scope of group codes, which may be used
+#if 0
```
Comment (RingsC) [inline storage/tianmu/core/value_matching_table.cpp]: If these codes are useless, pls, remove it now, not just comment it out.

File: storage/tianmu/core/value_matching_hashtable.cpp [hunk 6]
```diff
@@ -114,6 +114,49 @@ int64_t ValueMatching_HashTable::ByteSize() {
   return res + max_no_rows * total_width;
 }
 
+bool ValueMatching_HashTable::FindCurrentRow(unsigned char *input_buffer, int64_t &row, bool add_if_new,
```
Comment (RingsC) [inline storage/tianmu/core/value_matching_hashtable.cpp:117]: All new added codes should pass clang-format checking.

File: storage/tianmu/core/rc_attr.h [hunk 7]
```diff
@@ -123,6 +123,11 @@ class RCAttr final : public mm::TraceableObject, public PhysicalColumn, public P
     return dpn.min_i;
   }
 
+  Pack *GetPack(int64_t obj) {
```
Comment (RingsC) [inline storage/tianmu/core/rc_attr.h]: The `obj` is what? and if we can foud or some errors occur, what's value of `pack` ?

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 8]
```diff
@@ -234,9 +235,18 @@ void AggregationAlgorithm::MultiDimensionalGroupByScan(GroupByWrapper &gbw, int6
   }
   gbw.SetDistinctTuples(mit.NumOfTuples());
 
+  auto get_thd_cnt = []() {
```
Comment (RingsC) [inline storage/tianmu/core/aggregation_algorithm.cpp:238]: Why 4? not 5 or somthing else.

Comment (RingsC) [reply storage/tianmu/core/aggregation_algorithm.cpp:238]: and I think that you miss these 
```
if (ParallelAllowed(gbw) && !limit_less_than_no_groups) {
    thd_cnt = std::thread::hardware_concurrency() / 4; // For concurrence reason, don't swallow all
}
```
the original one has been stated the condition of parallelization.

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 9]
```diff
@@ -234,9 +235,18 @@ void AggregationAlgorithm::MultiDimensionalGroupByScan(GroupByWrapper &gbw, int6
...
+    int hardware_concurrency = std::thread::hardware_concurrency();
+    return hardware_concurrency > 4 ? (hardware_concurrency / 4) : 1;
+  };
+
   int thd_cnt = 1;
-  if (ParallelAllowed(gbw) && !limit_less_than_no_groups) {
-    thd_cnt = std::thread::hardware_concurrency() / 4;  // For concurrence reason, don't swallow all cores once.
+  if (force_parall) {
```
Comment (RingsC) [inline storage/tianmu/core/aggregation_algorithm.cpp:245]: I thik we can refine these comparsion statement to be more concise.

File: storage/tianmu/core/aggregation_algorithm.cpp [hunk 10]
```diff
@@ -361,7 +371,7 @@ void AggregationAlgorithm::MultiDimensionalGroupByScan(GroupByWrapper &gbw, int6
         displayed_no_groups = t->NumOfObj();
       if (t->NumOfObj() >= limit) break;
       if (gbw.AnyTuplesLeft()) gbw.ClearUsed();  // prepare for the next pass, if needed
-    } while (gbw.AnyTuplesLeft());               // do the next pass, if anything left
+    } while (gbw.AnyTuplesLeft() && (1 == thd_cnt));  // do the next pass, if anything left
```
Comment (RingsC) [inline storage/tianmu/core/aggregation_algorithm.cpp:375]: thd_cnt will update its value in every `while loop` or not?  which statement does update the value of `thd_cnt`?

**Review by RingsC — CHANGES_REQUESTED** [review]

File: storage/tianmu/vc/virtual_column_base.cpp (hunk 3 above)
Comment (adofsauron) [reply storage/tianmu/vc/virtual_column_base.cpp]: If do not do this, Q16 will not be executed

File: storage/tianmu/core/value_matching_table.h (hunk 4 above)
Comment (adofsauron) [reply storage/tianmu/core/value_matching_table.h:55]: In the subclass ValueMatching_HashTable, used to compare the lengths of matching strings

File: storage/tianmu/core/aggregation_algorithm.cpp (hunk 10 above)
Comment (adofsauron) [reply storage/tianmu/core/aggregation_algorithm.cpp:375]: no, never after be set value over

File: storage/tianmu/core/aggregation_algorithm.cpp (hunk 9 above)
Comment (adofsauron) [reply storage/tianmu/core/aggregation_algorithm.cpp:245]: yes, let me think about a minutes

File: storage/tianmu/core/rc_attr.h (hunk 7 above)
Comment (adofsauron) [reply storage/tianmu/core/rc_attr.h]: no more use this code, to delete

File: storage/tianmu/core/value_matching_hashtable.cpp (hunk 6 above)
Comment (adofsauron) [reply storage/tianmu/core/value_matching_hashtable.cpp:117]: that's great

File: storage/tianmu/core/value_matching_table.cpp (hunk 5 above)
Comment (adofsauron) [reply storage/tianmu/core/value_matching_table.cpp]: It is better to leave it, now that the entire code logic does not cover all the details, and to add a parameter control

File: storage/tianmu/core/value_matching_table.h (hunk 4 above)
Comment (adofsauron) [reply storage/tianmu/core/value_matching_table.h:55]: > 

It's better to keep readability and functional logic separate, or else the code that goes wrong and needs to be rolled back will be too big

**Review by isredstar — APPROVED** [review]
Nice job， but we hope from now on there will be stronger code for the multithread query processing.

**Review by hustjieke — APPROVED** [review]
LGTM

**Pull request merged**